### ISSUE TITLE
[BUG] "Cannot use import statement outside a module" when imported from another dependency in Next app

### ISSUE THREAD

**Issue by benface** [react-dropzone/react-dropzone#1377]
**Describe the bug**
This is a follow-up to #1259, and in fact [it was also reported](https://github.com/react-dropzone/react-dropzone/issues/1259#issuecomment-2392005987) by @CestDiego in that same issue but after it got closed. I thought I'd create a separate issue for it.

I am not sure why, but it looks like Next (latest version here, `14.2.14`) expects transitive dependencies (i.e. dependencies of dependencies) that have both CJS and ESM exports to have `"type": "module"` in their `package.json`, otherwise we get the following error:

```
Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
unhandledRejection
/Users/[my-project]/node_modules/.pnpm/react-dropzone@14.2.9_react@18.3.1/node_modules/react-dropzone/dist/es/index.js:37
import React, { forwardRef, Fragment, useCallback, useEffect, useImperativeHandle, useMemo, useReducer, useRef } from "react";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (node:internal/modules/cjs/loader:1378:20)
    at Module._compile (node:internal/modules/cjs/loader:1428:41)
    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12)
    at cjsLoader (node:internal/modules/esm/translators:346:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:286:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:234:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:473:24) {
  type: 'SyntaxError'
}
```

**To Reproduce**
https://codesandbox.io/p/devbox/next-react-dropzone-bug-mj3qdk
Notice I am importing from `local-package`, which itself is a ESM package that imports `react-dropzone`. That is what causes the issue; importing `react-dropzone` directly from the Next app seems to work.

**Expected behavior**
There should not be any error.

**Comment by benface** [comment]
Potentially related: https://github.com/vercel/next.js/issues/61715

**Comment by benface** [comment]
Ok so I believe this can be fixed by using the `.mjs` extension for the ESM exports, and making sure that all `import`s in ESM include the file's extension. Basically what @jacobclarke92 wrote [here](https://github.com/react-dropzone/react-dropzone/issues/1259#issuecomment-1404288766).

**Comment by LeahyCC** [comment]
Same issue on Remix project.

```
[vite] Error when evaluating SSR module /[redacted]/route.tsx: failed to import "react-dropzone"
|- /[redacted]/node_modules/react-dropzone/dist/es/index.js:37
import React, { forwardRef, Fragment, useCallback, useEffect, useImperativeHandle, useMemo, useReducer, useRef } from "react";
^^^^^^
```

**Comment by jakeymoreland** [comment]
> Same issue on Remix project.
> 
> ```
> [vite] Error when evaluating SSR module /[redacted]/route.tsx: failed to import "react-dropzone"
> |- /[redacted]/node_modules/react-dropzone/dist/es/index.js:37
> import React, { forwardRef, Fragment, useCallback, useEffect, useImperativeHandle, useMemo, useReducer, useRef } from "react";
> ^^^^^^
> ```

If you're not using vite and using remix.config add react-dropzone to serverDependenciesToBundle in remix.config

**Comment by rolandjitsu** [comment]
@benface does the latest patch work for you?

**Comment by benface** [comment]
No @rolandjitsu, same issue.

**Comment by rolandjitsu** [comment]
> No @rolandjitsu, same issue.

Hmm, it's probably because of the dependency (file-selector or attr-accept) or it may require the `"type": "module"` to be set or the use of `.mjs` in the file extensions. Do you think you could provide a repo or sandobox that reproduces this issue?

**Comment by benface** [comment]
@rolandjitsu – I did provide a CodeSandbox in my OP, but I just updated it to use the latest version. You can see that the error still occurs (at least it does when running `pnpm build` in the terminal).

**Comment by rolandjitsu** [comment]
@benface checkout [next-react-dropzone-bug-forked-6ll2th](https://codesandbox.io/p/devbox/next-react-dropzone-bug-forked-6ll2th?workspaceId=ccdfbac0-f3ef-49aa-a98e-b97259899e70). This seems to be working when importing this lib directly. I've also gone back in version for nextjs as the latest versions is throwing some error that I haven't had the time to look into.

**Comment by benface** [comment]
Thanks for the workarounds @rolandjitsu, really appreciate you looking into this. For my use case though, I have a (private) component library that uses the main `Dropzone` component, and that library is used in some Next apps, so it is still a problem, right? Have you looked into whether [this proposed solution](https://github.com/react-dropzone/react-dropzone/issues/1259#issuecomment-1404288766) (renaming the files to `.mjs`) would fix it?

**Comment by rolandjitsu** [comment]
> Thanks for the workarounds @rolandjitsu, really appreciate you looking into this. For my use case though, I have a (private) component library that uses the main `Dropzone` component, and that library is used in some Next apps, so it is still a problem, right? Have you looked into whether [this proposed solution](https://github.com/react-dropzone/react-dropzone/issues/1259#issuecomment-1404288766) (renaming the files to `.mjs`) would fix it?

If you checkout the 2nd sandbox, it imports from the `local-package`, so that should work. The caveat is that it seems to only work with the hook `useDropzone` and not with the jsx element. And the nextjs version is on v13. Those are probably unrelated issues.

**Comment by benface** [comment]
I don't think it has to do with the fact that it's a JSX element, but rather that it's the default export, which sounds related to ESM vs CJS to me. 🤔

**Comment by rolandjitsu** [comment]
> I don't think it has to do with the fact that it's a JSX element, but rather that it's the default export, which sounds related to ESM vs CJS to me. 🤔

Possibly. I'd need to look more into it.

**Comment by enBonnet** [comment]
If you are on Astrojs, this worked for me:

```
export default defineConfig({
  integrations: [tailwind(), react()],
  vite: {
    ssr: {
      noExternal: ["react-dropzone"],
    },
  },
});
```

The snippet gave by @jakeymoreland should work for all the apps sitting on Vitejs.

**Comment by benface** [comment]
This seems fixed in the latest version, I believe by https://github.com/react-dropzone/react-dropzone/commit/201687900724b45ec98d26cde3626a1c6687c9e1. Thank you!
